### PR TITLE
Add static method support to core Variant types

### PR DESCRIPTION
**Pull request by reduz**
* Properly exposed, including validated and variant call
* Bound static functions in String and Color
* Did not add support for scripting languages or GDNative, will have to be added manually by @vnen 

image

### CODE REVIEW

File: editor/doc_tools.cpp [hunk 1]
```diff
@@ -394,13 +394,22 @@ void DocTools::generate(bool p_basic_types) {
 					method.qualifiers += " ";
 				}
 				method.qualifiers += "const";
-			} else if (E->get().flags & METHOD_FLAG_VARARG) {
+			}
```
Comment (reduz) [inline editor/doc_tools.cpp:397]: This seemed wrong somehow, not quite sure why it was like this, so I changed it.

**Comment by KoBeWi** [comment]
Does it support non-variant classes like e.g. File?

**Comment by reduz** [comment]
@KoBeWi not for now, maybe in a future one. One thing at a time.

**Comment by akien-mga** [comment]
This error is shown when starting the project manager or editing any project:
```
ERROR: Condition "!imi.is_vararg && imi.argument_count != imi.argument_names.size()" is true.
   at: register_builtin_method (core/variant/variant_call.cpp:691)
```

Otherwise seems fine. A description will need to be added for the new `static` qualifier in:
https://github.com/godotengine/godot/blob/369dffc2179b41477a9baa566ed70b94d14b967a/doc/tools/makerst.py#L1050-L1056

(As a side note, it could be nice to have those abbr tooltips in the in-editor docs too @Calinou.)

**Comment by Calinou** [comment]
> (As a side note, it could be nice to have those abbr tooltips in the in-editor docs too @Calinou.)

This requires https://github.com/godotengine/godot-proposals/issues/1285 to be implemented first.

**Comment by vnen** [comment]
> This error is shown when starting the project manager or editing any project:
> 
> ```
> ERROR: Condition "!imi.is_vararg && imi.argument_count != imi.argument_names.size()" is true.
>    at: register_builtin_method (core/variant/variant_call.cpp:691)
> ```

This is because of `String::num`. It is bound with 3 arguments but the actual function only has 2.

File: core/variant/variant.h [hunk 2]
```diff
@@ -495,13 +495,16 @@ class Variant {
...
 	static bool is_builtin_method_const(Variant::Type p_type, const StringName &p_method);
+	static bool is_builtin_method_static(Variant::Type p_type, const StringName &p_method);
 	static bool is_builtin_method_vararg(Variant::Type p_type, const StringName &p_method);
 	static void get_builtin_method_list(Variant::Type p_type, List<StringName> *p_list);
 	static int get_builtin_method_count(Variant::Type p_type);
 
 	void call(const StringName &p_method, const Variant **p_args, int p_argcount, Variant &r_ret, Callable::CallError &r_error);
 	Variant call(const StringName &p_method, const Variant &p_arg1 = Variant(), const Variant &p_arg2 = Variant(), const Variant &p_arg3 = Variant(), const Variant &p_arg4 = Variant(), const Variant &p_arg5 = Variant());
 
+	void call_static(Variant::Type p_type, const StringName &p_method, const Variant **p_args, int p_argcount, Variant &r_ret, Callable::CallError &r_error);
```
Comment (vnen) [inline core/variant/variant.h:506]: Shouldn't this be static?

**Comment by vnen** [comment]
Also (not sure if you want to change this here), but `Color::from_hsv()` should be made static too.

**Comment by vnen** [comment]
I updated this to make the `call_static` function static and fixed the binding of `String::num` to match the signature.

For the `Color::from_hsv` and other instances (like whether we want the `String::num_int64` which allows to set the base) can be left for other PRs.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**